File: src/res_web.js
const _res = {
    readFile(_namePtr, _nameLen, _filePtr, _fileLen) {
        fetch(_utils._getString(_namePtr, _nameLen))
            .then(_response => _response.arrayBuffer())
            .then(_arrayBuffer => {
                _utils._u8Array(_filePtr, _fileLen).set(
                    new Uint8Array(_arrayBuffer)
                );
                _utils._getWasm().readFileComplete(true);
            })
            .catch(_err => {
                console.log(_err);
                _utils._getWasm().readFileComplete(false);
            });
    }
};

File: src/res_web.zig
const js = struct {
    extern fn readFile(
        name_ptr: [*]const u8,
        name_len: usize,
        file_ptr: [*]const u8,
        file_len: usize,
    ) void;
};

var read_file_frame: anyframe = undefined;
var read_file_result: anyerror!void = undefined;

pub fn readFile(name: []const u8, file: []u8) !void {
    try await async readFileAsync(name, file);
}

fn readFileAsync(name: []const u8, file: []u8) !void {
    js.readFile(name.ptr, name.len, file.ptr, file.len);
    suspend {
        read_file_frame = @frame();
    }
    try read_file_result;
}

export fn readFileComplete(success: bool) void {
    if (success) {
        read_file_result = {};
    } else {
        read_file_result = error.LoadFileFailed;
    }
    resume read_file_frame;
}
